bot/exts/utils/clean.py

```diff
@@ -106,69 +224,84 @@ def predicate_regex(message: Message) -> bool:
-        message_ids = []
         self.cleaning = True
 
-        # Find the IDs of the messages to delete. IDs are needed in order to ignore mod log events.
-        for channel in channels:
-            async for message in channel.history(limit=amount):
+        if use_cache:
+            message_mappings, message_ids = self._get_messages_from_cache(amount=amount, predicate=predicate)
+        else:
+            message_mappings, message_ids = await self._get_messages_from_channels(
+                amount=amount,
+                channels=channels,
+                predicate=predicate,
+                until_message=until_message
+            )
+
+        if not self.cleaning:
+            # Means that the cleaning was canceled
+            return
+
+        # Now let's delete the actual messages with purge.
+        self.mod_log.ignore(Event.message_delete, *message_ids)
+
+        for channel, messages in message_mappings.items():
+
+            to_delete = []
+
+            for current_index, message in enumerate(messages):
 
-                # If at any point the cancel command is invoked, we should stop.
                 if not self.cleaning:
+                    # Means that the cleaning was canceled
                     return
 
-                # If we are looking for specific message.
-                if until_message:
+                if self.is_older_than_14d(message):
+                    # further messages are too old to be deleted in bulk
+                    await self._delete_messages_individually(messages[current_index:])
+                    if not self.cleaning:
+                        # Means that deletion was canceled while deleting the individual messages
+                        return
+                    break
 
-                    # we could use ID's here however in case if the message we are looking for gets deleted,
-                    # we won't have a way to figure that out thus checking for datetime should be more reliable
-                    if message.created_at < until_message.created_at:
-                        # means we have found the message until which we were supposed to be deleting.
-                        break
+                to_delete.append(message)
 
-                    # Since we will be using `delete_messages` method of a TextChannel and we need message objects to
-                    # use it as well as to send logs we will start appending messages here instead adding them from
-                    # purge.
-                    messages.append(message)
+                if len(to_delete) == 100:
+                    # we can only delete up to 100 messages in a bulk
+                    await channel.delete_messages(to_delete)
+                    to_delete.clear()
 
-                # If the message passes predicate, let's save it.
-                if predicate is None or predicate(message):
-                    message_ids.append(message.id)
+            if len(to_delete) > 0:
```

Comment:
For whatever reason this isn't deleting messages when they come from the cache. I made sure that the code actually reaches that point, and it seems like `to_delete` is populated as expected.

Reply:
Could you explain how you produced this bug? I naively just spammed some messages, and then used clean user, and it worked as expected